controllers/imagebasedgroupupgrade.go

```diff
@@ -0,0 +1,264 @@
+	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	mwv1alpha1 "open-cluster-management.io/api/work/v1alpha1"
+)
+
+// IBGUReconciler reconciles a ImageBasedGroupUpgrade object
+type IBGUReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=lcm.openshift.io,resources=imagebasedgroupupgrades,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=lcm.openshift.io,resources=imagebasedgroupupgrades/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=lcm.openshift.io,resources=imagebasedgroupupgrades/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *IBGUReconciler) Reconcile(ctx context.Context, req ctrl.Request) (nextReconcile ctrl.Result, err error) {
+	r.Log.Info("Start reconciling IBGU", "name", req.NamespacedName)
+	defer func() {
+		if nextReconcile.RequeueAfter > 0 {
+			r.Log.Info("Finish reconciling IBGU", "name", req.NamespacedName, "requeueAfter", nextReconcile.RequeueAfter.Seconds())
+		} else {
+			r.Log.Info("Finish reconciling IBGU", "name", req.NamespacedName, "requeueRightAway", nextReconcile.Requeue)
+		}
+	}()
+
+	nextReconcile = doNotRequeue()
+
+	ibgu := &ibguv1alpha1.ImageBasedGroupUpgrade{}
+	err = r.Get(ctx, req.NamespacedName, ibgu)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			err = nil
+			return
+		}
+		r.Log.Error(err, "Failed to get IBGU")
+		return
+	}
+	err = r.ensureManifests(ctx, ibgu)
+	if err != nil {
+		r.Log.Error(err, "error ensure manifests")
+	}
+	utils.SetStatusCondition(&ibgu.Status.Conditions,
+		utils.ConditionTypes.ManifestsCreated,
+		utils.ConditionReasons.Completed,
+		metav1.ConditionTrue,
+		"All manifests are created")
+
+	err = r.syncStatusWithCGUs(ctx, ibgu)
+	if err != nil {
+		r.Log.Error(err, "error syncing status with CGUs")
+	}
+
+	// Update status
+	err = r.updateStatus(ctx, ibgu)
+	return
+}
+
+func (r *IBGUReconciler) syncStatusWithCGUs(ctx context.Context, ibgu *ibguv1alpha1.ImageBasedGroupUpgrade) error {
+	cguList := &ranv1alpha1.ClusterGroupUpgradeList{}
+	err := r.List(ctx, cguList, &client.ListOptions{
+		Namespace:     ibgu.Namespace,
+		LabelSelector: labels.SelectorFromSet(labels.Set{utils.CGUOwnerIBGULabel: ibgu.Name}),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get CGUs for the IBGU: %w", err)
+	}
+	if len(cguList.Items) == 0 {
+		return nil
+	}
+	m := make(map[string]ibguv1alpha1.ClusterState)
+	for _, cgu := range cguList.Items {
```

Comment:
The CGUs should be sorted by the order of the actions. Also, we probably can ignore completed CGUs?

Reply:
Hmm, I guess we can't skip completed CGUs if they are chained rather than getting enabled by this controller.